### DISCUSSION TITLE
How to properly integrate Angular applications into Eclipse Theia?

### DISCUSSION THREAD

**Discussion by friendlyAce** [eclipse-theia/theia#13723] [Show and tell]
Hi, 
i recently discovered this project and it seems pretty exiting and IMO has a lot of potential.
We are currently using the Eclipse RCP framwork for creating a custom workbench with plenty of views etc.
We are using an custom chromium embedded java-cef (Java SWT port of cef) build to host an angular application within jcef in an eclipse editor.
We still have some heavy dependencies to Eclipse/RCP (project/user Preferences, osgi services, selection handling, eclipse and custom extension points, etc.)
I havent read the full Theia documentation yet, but from what i was already reading, this shouldnt be a big deal, since most of these features are available in theia, just have to be re-implemented/migrated.
Ill need to investigate and do some testing on that later on.

My objective:
Currently, i am trying to create a proof of concept for our project on how the migration away from Eclipse RCP to Eclipse Theia could work,
so that we can consider migrating our project to it, as our development is more web-focused anyways and for most parts we dont heavily rely on SWT etc. anymore.

My aim is to embed an Angular 17 application within the editor section of Eclipse Theia.
However, I couldn't find comprehensive examples or documentation for this specific use case, except for the React widget example or the Tree editor within theia.
And when searching for "Eclipse Theia and Angular", there its only mentioned that Eclipse theia just goes well with web frameworks such as React, Vue, Angular in general, but i couldnt find much more detail, docs, examples on that.

So I was wondering what the preferred way for embedding an Angular 17 application as an editor/widget in the ecipse theia framework should look like and how the angular app can communicate with the extensions/features of the theia framework.
From what i was reading in the Issues when searching for the keyword Angular, is that most people asked for a way to integrate Theia as module/section within existing angular applications, via the core theia npm dependency.
I also tried that, havent had much success to get that running yet, but this doesnt really seem relevant to me for the project, as we want to use Theia as a RCP framework replacement to display angular developed components/logic and not the other way around, so i discarded that approach.

Proof of Concept:
To familiarise with theia and explore the integration a little bit and for getting some insights on how the concepts and architecture in Theia works, i created a small PoC project, the sources can be found here: https://github.com/friendlyAce/theia-application
I discovered and adapted the React widget example for my angular application.

I first started off by creating a simple angular 17 demo project.
https://github.com/friendlyAce/theia-application/tree/master/theia-angular-app

After that, with yo theia-extension i created a simple hello world and editor extension, to see how the workflow with yarn works on a multi npm project mono repo and how i could build & run both the electron desktop app and the browser app.
After doing some testing and trying out the example, on how to add context menu entries to existing menus and how to register the contributor & consumer logic i started working on the PoC.

My first approach was to extend the Theia Base Widget, https://github.com/friendlyAce/theia-application/blob/master/theia-extensions/angular-widget/src/browser/angular-widget/angular-root-widget.ts
that internally used an iframe to reference the running angular application on port 4200 via <iframe src="http://localhost:4200">, this worked fine, but this requires the angular app to be served/hosted separately.

After a bit of trying and since angular apps are just some bundled html and js, i thought of maybe i can use the angular build dist/ output of my angular app and make Theia access, execute & instanizate the root app component.

For that i extended my AngularWidget, to just create some div container that later will hold my app-root angular entry component and add the main.js script of my angular app to the script headers.
Now all there was left to somehow figure out where to put the angular dist/ build output and copy-paste it there via some npm/yarn script, to both the electron & the browser directories.

After fixing some smaller issues, i ended up seeing my running angular app in a view in eclipse theia:
Screenshot: https://github.com/friendlyAce/theia-application/blob/master/poc_screenshots/electron_and_browser_with_ng.png

However, this approach somehow doesn't feel quite right, more like a kind of workaround, which is why I wanted to ask how such an integration of a web framework (like explicit angular) should actually look like?

Questions
- Preferred Method: What is the preferred way to embed an Angular application as an editor/widget within Eclipse Theia? Are there any recommended best practices or patterns for this integration?
- Documentation/Examples: Are there any existing examples or detailed documentation on embedding non-React applications (specifically Angular) within Theia? If not, would the team consider creating or pointing to a community-driven example?
- Serving and Bundling: Is there a more seamless way to bundle and serve the Angular application within Theia without requiring separate hosting or extensive manual copying of files? To also make hot reloading possible, for ease of developing?
- Common Pitfalls: Are there any common pitfalls or challenges that we should be aware of when embedding an Angular application within Theia?
- Performance Considerations: How does the performance of embedding an Angular application within Theia compare to the RCP approach? Are there any known performance bottlenecks, does Theia perform better than RCP?
- Debugging and Tooling: What tools and strategies are available for debugging Angular applications embedded in Theia? Are there specific configurations needed to enable efficient debugging?
- Inter-Component Communication: Previously we were using Chromiums interapp communication to communicate via our angular FE & java/eclipse RCP backend, What are the best practices for communication between the Angular application and other Theia components or Theia framework extensions/features? Are there recommended patterns for event handling and data sharing?
- Integration with Theia Services: How can the Angular application leverage Theia’s built-in services and APIs (e.g., file system, command registry)? Are there examples or documentation on integrating these services?
- Packaging and Deployment: What is the recommended approach for packaging and deploying a Theia application that includes an embedded Angular application? Are there specific deployment strategies to simplify this process?

BR,
Vincent

**Comment by msujew** [comment]
Hey @friendlyAce,

not-an-angular dev here, so don't expect too much input on the Angular front from me, but I can answer some of your other questions.

> What is the preferred way to embed an Angular application as an editor/widget within Eclipse Theia? Are there any recommended best practices or patterns for this integration?

Generally, you get the best UX when you directly integrate your app as a widget. Something like the [Angular elements](https://angular.io/guide/elements) guide should result in good integration. Once you have a web-component for your Angular element, you should be able to render it into a Theia `BaseWidget` (pure HTML widget).

> Are there any existing examples or detailed documentation on embedding non-React applications (specifically Angular) within Theia? If not, would the team consider creating or pointing to a community-driven example?

non-React, yes. since Theia is mainly pure HTML. It's just some components that are written in React, so we offer some integration for that. I'm not aware of any public examples that offer anything wrt to that.

> Serving and Bundling: Is there a more seamless way to bundle and serve the Angular application within Theia without requiring separate hosting or extensive manual copying of files?

If you follow the guide linked above, webpack should bundle everything into a single (chunked) bundle. That should alleviate the need to provide separate hosting for the angular app.

>To also make hot reloading possible, for ease of developing?

There currently is no hot reloading available, since Theia isn't served by webpack, but by its own backend server. I'm not sure how well hot reloading can be integrated into the backend.

> Common Pitfalls: Are there any common pitfalls or challenges that we should be aware of when embedding an Angular application within Theia?

None that I'm aware of.

> Performance Considerations: How does the performance of embedding an Angular application within Theia compare to the RCP approach? Are there any known performance bottlenecks, does Theia perform better than RCP?

I'm not sure how well RCP-over-SWT/CEF performs, but Theia generally behaves quite well due to its decentralized rendering model - most work is done on the backend and the plugin host (a separate process), while the frontend is completely separated from that. Due to that, the frontend usually runs fairly smooth, even if the other processes are doing some heavy lifting.

> Debugging and Tooling: What tools and strategies are available for debugging Angular applications embedded in Theia? Are there specific configurations needed to enable efficient debugging?

None that I'm aware of.

> Inter-Component Communication: Previously we were using Chromiums interapp communication to communicate via our angular FE & java/eclipse RCP backend, What are the best practices for communication between the Angular application and other Theia components or Theia framework extensions/features? Are there recommended patterns for event handling and data sharing?

You would just integrate your Angular code directly with the rest of the JavaScript based frontend application. Theia features a bunch of event handling mechanisms based on an emitter/event pattern. The communication with the backend is done via RPC (remote procedure call), see [here](https://theia-ide.org/docs/json_rpc/) for more info.

> Integration with Theia Services: How can the Angular application leverage Theia’s built-in services and APIs (e.g., file system, command registry)? Are there examples or documentation on integrating these services?

We just use inversify, a dependency injection library to integrate with services. See [this guide](https://theia-ide.org/docs/services_and_contributions/).

> Packaging and Deployment: What is the recommended approach for packaging and deploying a Theia application that includes an embedded Angular application? Are there specific deployment strategies to simplify this process?

Since the angular app should be part of the bundle, there shouldn't be any additional packaging/deployment process necessary.

**Reply by sdirix** [reply]
@msujew That is a great answer.

There is one more integration approach to consider: Assuming that your Angular application is mostly self-contained you could also integrate it as a web view via a VS Code extension. This also has the benefit that it could run within VS Code itself, in case that is a use case for you. With a bit of tinkering you can even setup hot reloading there.

If the application is deeply integrated, this approach becomes less useful as you will need to setup a lot of communication. If you intend to run singular Angular components within Theia widgets then this approach does not work for you. I would then recommend using the webcomponent approach instead.
